plonky2x/src/frontend/vars/array.rs

```diff
@@ -162,18 +170,75 @@ impl<L: PlonkParameters<D>, const D: usize> CircuitBuilder<L, D> {
         V::from_variables_unsafe(&selected_vars)
     }
 
-    pub fn get_fixed_subarray<const SIZE: usize>(
+    pub fn get_fixed_subarray<const MAX_ARRAY_SIZE: usize, const SUB_ARRAY_SIZE: usize>(
         &mut self,
-        array: &[Variable],
-        start: Variable,
-    ) -> ArrayVariable<Variable, SIZE> {
-        let mut res = Vec::new();
-        for i in 0..SIZE {
-            let i_variable = self.constant(L::Field::from_canonical_usize(i));
-            let index = self.add(start, i_variable);
-            res.push(self.select_array(array, index));
+        array: &ArrayVariable<Variable, MAX_ARRAY_SIZE>,
+        array_size: Variable,
+        start_idx: Variable,
+        seed: &ArrayVariable<ByteVariable, 15>, // TODO: Seed it with 120 bits.  Need to figure out if this is secure.
+    ) -> ArrayVariable<Variable, SUB_ARRAY_SIZE> {
+        // TODO:  Need to add check that array_size is less than MAX_ARRAY_SIZE.
+        // TODO:  Need to add check that start_idx + SUB_ARRAY_SIZE is less than array_size.
+
+        let sub_array = self.init::<ArrayVariable<Variable, SUB_ARRAY_SIZE>>();
+
+        self.add_simple_generator(SubArrayExtractor::<L, D, MAX_ARRAY_SIZE, SUB_ARRAY_SIZE> {
+            array: array.clone(),
+            array_size,
+            start_idx,
+            sub_array: sub_array.clone(),
+            _marker: PhantomData,
+        });
+
+        let mut challenger = RecursiveChallenger::<L::Field, PoseidonHash, D>::new(&mut self.api);
+        for seed_chunk in seed.as_vec().chunks(5) {
+            let seed_element_bits = seed_chunk
+                .iter()
+                .flat_map(|x| x.as_bool_targets())
+                .collect_vec();
+            let seed_element = self.api.le_sum(seed_element_bits.iter());
+            challenger.observe_element(seed_element);
+        }
+
+        for _i in 0..2 {
+            let challenges = challenger
+                .get_n_challenges(&mut self.api, SUB_ARRAY_SIZE)
+                .iter()
+                .map(|x| Variable::from(*x))
+                .collect_vec();
+            let sub_array_size = self.constant(L::Field::from_canonical_usize(SUB_ARRAY_SIZE));
+            let end_idx = self.add(start_idx, sub_array_size);
+            let mut within_sub_array = self.zero::<Variable>();
+            let one = self.one();
+
+            let mut accumulator1 = self.zero::<Variable>();
+            let mut j_target = self.zero();
+            for j in 0..MAX_ARRAY_SIZE {
+                let at_start_idx = self.is_equal(j_target, start_idx);
+                within_sub_array = within_sub_array.add(at_start_idx.variables()[0], self);
+                let at_end_idx = self.is_equal(j_target, end_idx);
+                within_sub_array = within_sub_array.sub(at_end_idx.variables()[0], self);
+
+                let mut subarray_idx = j_target.sub(start_idx, self);
```

Comment:
Imo we should just have a modular function that gets an array commitment and then build this code on top of that

Reply:
Do you mean a subarray commitment?